README.pod
=pod

=encoding UTF-8

=head1 NAME

MooseX::UndefTolerant - Make your attribute(s) tolerant to undef initialization

=head1 VERSION

version 0.21

=head1 SYNOPSIS

  package My::Class;

  use Moose;
  use MooseX::UndefTolerant;

  has 'name' => (
    is => 'ro',
    isa => 'Str',
    predicate => 'has_name'
  );

  # Meanwhile, under the city...

  # Doesn't explode
  my $class = My::Class->new(name => undef);
  $class->has_name # False!

Or, if you only want one attribute to have this behaviour:

  package My:Class;
  use Moose;

  use MooseX::UndefTolerant::Attribute;

  has 'bar' => (
      traits => [ qw(MooseX::UndefTolerant::Attribute)],
      is => 'ro',
      isa => 'Num',
      predicate => 'has_bar'
  );

=head1 DESCRIPTION

Loading this module in your L<Moose> class makes initialization of your
attributes tolerant of undef.  If you specify the value of undef to any of
the attributes they will not be initialized, effectively behaving as if you
had not provided a value at all.

You can also apply the 'UndefTolerant' trait to individual attributes. See
L<MooseX::UndefTolerant::Attribute> for details.

There will be no change in behaviour to any attribute with a type constraint
that accepts undef values (for example C<Maybe> types), as it is presumed that
since the type is already "undef tolerant", there is no need to avoid
initializing the attribute value with C<undef>.

As of Moose 1.9900, this module can also be used in a role, in which case all
of that role's attributes will be undef-tolerant.

=head1 MOTIVATION

I often found myself in this quandary:

  package My:Class;
  use Moose;

  has 'foo' => (
    is => 'ro',
    isa => 'Str',
  );

  # ... then

  my $foo = ... # get the param from something

  my $class = My:Class->new(foo => $foo, bar => 123);

What if foo is undefined?  I didn't want to change my attribute to be
Maybe[Str] and I still want my predicate (C<has_foo>) to work.  The only
real solution was:

  if(defined($foo)) {
    $class = My:Class->new(foo => $foo, bar => 123);
  } else {
    $class = My:Class->new(bar => 123);
  }

=for stopwords codemulch

Or some type of codemulch using ternary conditionals.  This module allows you
to make your attributes more tolerant of undef so that you can keep the first
example: have your cake and eat it too!

=head1 PER ATTRIBUTE

See L<MooseX::UndefTolerant::Attribute>.

=head1 CAVEATS

This extension does not currently work in immutable classes when applying the
trait to some (but not all) attributes in the class. This is because the
inlined constructor initialization code currently lives in
L<Moose::Meta::Class>, not L<Moose::Meta::Attribute>. The good news is that
this is expected to be changing shortly.

=head1 ACKNOWLEDGEMENTS

Many thanks to the crew in #moose who talked me through this module:

=for stopwords Doran Eldridge Hardison Pearcey diz dylan jshirley

=over 4

Hans Dieter Pearcey (confound)
Jesse Luehrs (doy)
Tomas Doran (t0m)
Dylan Hardison (dylan)
Jay Shirley (jshirley)
Mike Eldridge (diz)

=back

=head1 SUPPORT

Bugs may be submitted through L<the RT bug tracker|https://rt.cpan.org/Public/Dist/Display.html?Name=MooseX-UndefTolerant>
(or L<bug-MooseX-UndefTolerant@rt.cpan.org|mailto:bug-MooseX-UndefTolerant@rt.cpan.org>).

There is also a mailing list available for users of this distribution, at
L<http://lists.perl.org/list/moose.html>.

There is also an irc channel available for users of this distribution, at
L<C<#moose> on C<irc.perl.org>|irc://irc.perl.org/#moose>.

=head1 AUTHOR

Cory G Watson <gphat at cpan.org>

=head1 CONTRIBUTORS

=for stopwords Karen Etheridge Dave Rolsky Cory G Watson Jesse Luehrs Chris Andrews

=over 4

=item *

Karen Etheridge <ether@cpan.org>

=item *

Dave Rolsky <autarch@urth.org>

=item *

Cory G Watson <gphat@cpan.org>

=item *

Jesse Luehrs <doy@tozt.net>

=item *

Chris Andrews <chrisandrews@venda.com>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Cory G Watson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for MooseX-UndefTolerant

{{$NEXT}}

0.21      2016-10-23 01:05:18Z
   * avoid failing tests when upstream modules warn

0.20      2015-08-16 02:45:25Z
   * repository migrated to the github moose organization

0.19      2012-09-26 00:16:40Z
   * fix incorrect version dependency on Test::NoWarnings

0.18      2012-09-23 20:35:06Z
   * avoid undef warning by not attempting to manipulate constructor data for
     init_arg => undef attributes (thanks Damon Tkoch for the report!)

0.17      2012-05-03 21:43:42Z
    * re-release to fix dependency issue *again*! :)

0.16      2012-05-02 22:13:01Z
    * remove temporary dep, and increase the required version of
      Dist::Zilla::Plugin::Test::CheckDeps, that now injects this dependency
      automatically

0.15      2012-05-02 19:56:59Z
    * temporarily add missing dep on Test::CheckDeps

0.14      2012-05-02 16:23:54Z
    * re-release of 0.13 to fix a PAUSE indexing issue.

0.13    2012-03-25
    * unit test improvements - tighter TODO scopes, better diagnostics on
      failure, Test::Moose for cleaner logic
    * fix documentation referencing Moose changes that are still TBD
    * undef attr values being stripped at construction time are now removed
      from visibility of all attr initializations, not just the one being
      updated at the time
    * fixed case where an attribute on an immutable class was being stripped
      of its undef value at construction time even if its type constraint
      already can tolerate undef.

0.12    2011-04-03
    * This module can now be used in roles with Moose 1.9900+. (Jesse Luehrs)

0.11    2011-03-19
    * re-release to fix careless syntax error (whoops!) :)

0.10    2011-03-18
    * No longer apply UndefTolerant behaviour to attributes with a type
      constraint that handles undef (or those with no type constraint at all).
      (Karen Etheridge)

0.09    2011-03-02
    * Add missing test prereq on Test::Fatal. (Dave Rolsky)

0.08    2011-02-27
    * Additional (TODO) unit tests for immutable classes, and documentation
      notes about these limitations. (Karen Etheridge)

    * Added forward compatibility with Moose 1.9900+, but this module still
      works with older Moose too. (Dave Rolsky)

0.07    2010-10-21
    * Don't do extra work by exists-ing before defined-ing (Chip Salzenberg)

0.06    2010-10-11
    * Add support for immutable classes (Chris Andrews)
    * Be more careful avoiding warnings when the init_arg is undef (Chris Andrews)
    * POD fixups

0.05    2010-07-19
    * Avoid warnings from Moose 1.09. (Dave Rolsky)

0.04    2010-01-23
    * Rather than skipping initialization all together, just pop off the undef
      and let the init happen normally.  This makes default values work
      properly. (Test provided by (and thanks to) Tom Heady)

0.03    2009-12-03
    * Fiddle with dependencies.

0.02    2009-12-02
    * Actually check the value properly, therefore setting it properly

0.01    2009-12-01
        First version, released on an unsuspecting world.



lib/MooseX/UndefTolerant/Class.pm
package MooseX::UndefTolerant::Class;

our $VERSION = '0.22';

# applied to metaclass, for Moose >= 1.9900

use strict;
use warnings;

use Moose::Role;

# TODO: this code should be in the attribute trait, in the inlined version of
# initialize_instance_slot, but this does not yet exist!

around _inline_init_attr_from_constructor => sub {
    my $orig = shift;
    my $self = shift;
    my ($attr, $idx) = @_;

    my @source = $self->$orig(@_);

    my $init_arg = $attr->init_arg;
    my $type_constraint = $attr->type_constraint;
    my $tc_says_clean = ($type_constraint && !$type_constraint->check(undef) ? 1 : 0);

    # FIXME: not properly sanitizing field names - e.g. consider a field name "Z'ha'dum"
    return ($tc_says_clean ? (
        "if ( exists \$params->{'$init_arg'} && defined \$params->{'$init_arg'} ) {",
        ) : (),
        @source,
        $tc_says_clean ? (
        '} else {',
            "delete \$params->{'$init_arg'};",
        '}',
        ) : (),
    );
};

no Moose::Role;
1;


lib/MooseX/UndefTolerant/Attribute.pm
package MooseX::UndefTolerant::Attribute;

our $VERSION = '0.22';

use Moose::Role;
use namespace::autoclean;

around('initialize_instance_slot', sub {
    my $orig = shift;
    my $self = shift;
    my ($meta_instance, $instance, $params) = @_;

    my $key_name = $self->init_arg;

    # If our parameter passed in was undef, remove it from the parameter list...
    # but leave the value unscathed if the attribute's type constraint can
    # handle undef (or doesn't have one, which implicitly means it can)
    if (defined $key_name and not defined($params->{$key_name}))
    {
        my $type_constraint = $self->type_constraint;
        if ($type_constraint and not $type_constraint->check(undef))
        {
            delete $params->{$key_name};
        }
    }

    # Invoke the real init, as the above line cleared the undef param value
    $self->$orig(@_)
});

1;

# ABSTRACT: Make your attribute(s) tolerant to undef initialization

__END__

=head1 SYNOPSIS

  package My:Class;
  use Moose;

  use MooseX::UndefTolerant::Attribute;

  has 'bar' => (
      traits => [ qw(MooseX::UndefTolerant::Attribute)],
      is => 'ro',
      isa => 'Num',
      predicate => 'has_bar'
  );

  # Meanwhile, under the city...

  # Doesn't explode
  my $class = My::Class->new(bar => undef);
  $class->has_bar # False!

=head1 DESCRIPTION

Applying this trait to your attribute makes its initialization tolerant of
of undef.  If you specify the value of undef to any of the attributes they
will not be initialized (or will be set to the default, if applicable).
Effectively behaving as if you had not provided a value at all.

=cut


lib/MooseX/UndefTolerant/Constructor.pm
package MooseX::UndefTolerant::Constructor;

our $VERSION = '0.22';

# applied to constructor method metaclass, for Moose < 1.9900

use Moose::Role;

use strict;
use warnings;

around _generate_slot_initializer => sub {
    my $orig = shift;
    my $self = shift;

    # note the key in the params may not match the attr name.
    my $key_name = $self->_attributes->[$_[0]]->init_arg;

    # insert a line of code at the start of the initializer,
    # clearing the param if it's undefined.

    if (defined $key_name)
    {
        # leave the value unscathed if the attribute's type constraint can
        # handle undef (or doesn't have one, which implicitly means it can)
        my $type_constraint = $self->_attributes->[$_[0]]->type_constraint;
        if ($type_constraint and not $type_constraint->check(undef))
        {
            my $tolerant_code =
                qq# delete \$params->{'$key_name'} unless # .
                qq# exists \$params->{'$key_name'} && defined \$params->{'$key_name'};\n#;

            return $tolerant_code . $self->$orig(@_);
        }
    }

    return $self->$orig(@_);
};

no Moose::Role;
1;


lib/MooseX/UndefTolerant.pm
package MooseX::UndefTolerant;

our $VERSION = '0.22';

use strict;
use warnings;

use Moose 0.89 qw();
use Moose::Exporter;

use MooseX::UndefTolerant::Attribute;
use MooseX::UndefTolerant::Class;
use MooseX::UndefTolerant::Constructor;


my %metaroles = (
    class_metaroles => {
        attribute => [ 'MooseX::UndefTolerant::Attribute' ],
    }
);
if ( $Moose::VERSION < 1.9900 ) {
    $metaroles{class_metaroles}{constructor} = [
        'MooseX::UndefTolerant::Constructor',
    ];
}
else {
    $metaroles{class_metaroles}{class} = [
        'MooseX::UndefTolerant::Class',
    ];
    $metaroles{role_metaroles} = {
        applied_attribute => [
            'MooseX::UndefTolerant::Attribute',
        ],
        role => [
            'MooseX::UndefTolerant::Role',
        ],
        application_to_class => [
            'MooseX::UndefTolerant::ApplicationToClass',
        ],
        application_to_role => [
            'MooseX::UndefTolerant::ApplicationToRole',
        ],
    };
}


Moose::Exporter->setup_import_methods(%metaroles);

1;

# ABSTRACT: Make your attribute(s) tolerant to undef initialization

__END__

=head1 SYNOPSIS

  package My::Class;

  use Moose;
  use MooseX::UndefTolerant;

  has 'name' => (
    is => 'ro',
    isa => 'Str',
    predicate => 'has_name'
  );

  # Meanwhile, under the city...

  # Doesn't explode
  my $class = My::Class->new(name => undef);
  $class->has_name # False!

Or, if you only want one attribute to have this behaviour:

  package My:Class;
  use Moose;

  use MooseX::UndefTolerant::Attribute;

  has 'bar' => (
      traits => [ qw(MooseX::UndefTolerant::Attribute)],
      is => 'ro',
      isa => 'Num',
      predicate => 'has_bar'
  );

=head1 DESCRIPTION

Loading this module in your L<Moose> class makes initialization of your
attributes tolerant of undef.  If you specify the value of undef to any of
the attributes they will not be initialized, effectively behaving as if you
had not provided a value at all.

You can also apply the 'UndefTolerant' trait to individual attributes. See
L<MooseX::UndefTolerant::Attribute> for details.

There will be no change in behaviour to any attribute with a type constraint
that accepts undef values (for example C<Maybe> types), as it is presumed that
since the type is already "undef tolerant", there is no need to avoid
initializing the attribute value with C<undef>.

As of Moose 1.9900, this module can also be used in a role, in which case all
of that role's attributes will be undef-tolerant.

=head1 MOTIVATION

I often found myself in this quandary:

  package My:Class;
  use Moose;

  has 'foo' => (
    is => 'ro',
    isa => 'Str',
  );

  # ... then

  my $foo = ... # get the param from something

  my $class = My:Class->new(foo => $foo, bar => 123);

What if foo is undefined?  I didn't want to change my attribute to be
Maybe[Str] and I still want my predicate (C<has_foo>) to work.  The only
real solution was:

  if(defined($foo)) {
    $class = My:Class->new(foo => $foo, bar => 123);
  } else {
    $class = My:Class->new(bar => 123);
  }

=for stopwords codemulch

Or some type of codemulch using ternary conditionals.  This module allows you
to make your attributes more tolerant of undef so that you can keep the first
example: have your cake and eat it too!

=head1 PER ATTRIBUTE

See L<MooseX::UndefTolerant::Attribute>.

=head1 CAVEATS

This extension does not currently work in immutable classes when applying the
trait to some (but not all) attributes in the class. This is because the
inlined constructor initialization code currently lives in
L<Moose::Meta::Class>, not L<Moose::Meta::Attribute>. The good news is that
this is expected to be changing shortly.

=head1 ACKNOWLEDGEMENTS

Many thanks to the crew in #moose who talked me through this module:

=for stopwords Doran Eldridge Hardison Pearcey diz dylan jshirley

=for :list
Hans Dieter Pearcey (confound)
Jesse Luehrs (doy)
Tomas Doran (t0m)
Dylan Hardison (dylan)
Jay Shirley (jshirley)
Mike Eldridge (diz)

=cut
